### PR TITLE
Remove Closure Library dependency

### PR DESCRIPTION
**Pull request by JamiesWhiteShirt**
This is a follow-up to #173. It is also required for #172.

Openlayers removed this dependency because the Closure Library cannot be provided with ES modules. Therefore, if ol3-google-maps should provide ES2015 modules it cannot depend on the Closure Library either.

This has a some implications.

As you can see in the diff, `goog.asserts` and `goog.events` can no longer be used. Because Openlayers does not export `ol.asserts` and `ol.events`, I had to create equivalent replacements `olgm.asserts` and `olgm.events`.

Most instanceof assertions are now only typecasts. The remaining instanceof assertions are for checking user (developer) input.

I don't know how to avoid regressions on this. It seems there is no way to fully remove the dependency. Openlayers seems to have the same problem.

### CODE REVIEW

**Comment by adube** [comment]
@JamiesWhiteShirt You can rebase onto master.

**Comment by JamiesWhiteShirt** [comment]
I rebased this immediately after you asked, figured you might have missed it @adube

**Comment by adube** [comment]
Oh, yeah.  I didn't get a notice.  Don't hesitate to add a comment with @adube in the future right after such actions.  Sorry about being idle for so long.  I'll take care of reviewing this shortly.

File: src/olgm/obj.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Polyfill for Object.assign().  Assigns enumerable and own properties from
+ * one or more source objects to a target object.
+ *
+ * @see https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign
+ * @param {!Object} target The target object.
+ * @param {...Object} var_sources The source object(s).
+ * @return {!Object} The modified target object.
+ */
+olgm.obj.assign = (typeof Object.assign === 'function') ? Object.assign : function(target, var_sources) {
```
Comment (adube) [inline src/olgm/obj.js:14]: Would it have been possible to use the equivalent method from OpenLayers instead?

File: src/olgm/obj.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+
+/**
+ * Get an array of property values from an object.
+ * @param {Object<K,V>} object The object from which to get the values.
+ * @return {!Array<V>} The property values.
+ * @template K,V
+ */
+olgm.obj.getValues = function(object) {
```
Comment (adube) [inline src/olgm/obj.js:51]: Let's rename this one `values` and use the native `Object.values` if available.  See: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_objects/Object/values

File: src/olgm/obj.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  return values;
+};
+
+
+/**
+ * Determine if an object has any properties.
+ * @param {Object} object The object to check.
+ * @return {boolean} The object is empty.
+ */
+olgm.obj.isEmpty = function(object) {
```
Comment (adube) [inline src/olgm/obj.js:65]: same: use `ol.obj.isEmpty` instead?

File: src/olgm/obj.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  }
+  return output;
+};
+
+
+/**
+ * Removes all properties from an object.
+ * @param {Object} object The object to clear.
+ */
+olgm.obj.clear = function(object) {
```
Comment (adube) [inline src/olgm/obj.js:38]: same: use `ol.obj.clear` instead?

File: src/olgm/events.js [hunk 5]
```diff
@@ -0,0 +1,241 @@
+/* Based on https://github.com/openlayers/openlayers/blob/master/src/ol/events.js */
+goog.provide('olgm.events');
```
Comment (adube) [inline src/olgm/events.js:2]: Would it have been possible to use ol navite methods instead, that are the equivalent?

**Review by adube — COMMENTED** [review]
Great job!  I left some minor comments that should be addressed before merging, but everything is very clean. Please, tell me what you think.

**Comment by adube** [comment]
@JamiesWhiteShirt - just in case you hadn't noticed my comments.

**Comment by JamiesWhiteShirt** [comment]
Yeah, sorry, haven't had time. I'll get to it eventually.

**Comment by adube** [comment]
No problem.  Thanks for the reply and have a good one.

**Comment by JamiesWhiteShirt** [comment]
Hey @adube 

The reason we would initially provide our own `olgm.events.listen` is that the function we need is not exported by OpenLayers. This means OLGM will only work with raw access to OL, either in a full OL+OLGM bundle or by serving both OL and OLGM as libraries in closure-util. Using non-exported parts of OL broke the examples, because they use the OL bundle and OLGM added as a library in closure-util.

I have now modified the examples to serve both OL and OLGM as libraries in `serve.js`. The alternative would be to have the examples use an OL+OLGM bundle.

We could also try making a PR to add `ol.events.listen` to the OpenLayers API to avoid the trouble. I believe there is a minor risk of unexpected breaking changes to non-exported methods anyway.

**Comment by adube** [comment]
Ok, now I get it.  It makes sense.

>We could also try making a PR to add ol.events.listen to the OpenLayers API to avoid the trouble. I believe there is a minor risk of unexpected breaking changes to non-exported methods anyway.

No need to.  I agree with your addition of `olgm.events.listen`.

File: src/olgm/obj.js (hunk 2 above)
Comment (JamiesWhiteShirt) [reply src/olgm/obj.js:51]: `Object.values` and this function does actually have different behaviour. This function returns all values including prototype values, whereas `Object.values` excludes prototype values.

**Comment by JamiesWhiteShirt** [comment]
@adube I have reverted back to the addition of olgm.events.

**Comment by adube** [comment]
Excellent.

Would you please take care of the issue reported by Travis?

File: src/olgm/herald/feature.js [hunk 6]
```diff
@@ -207,13 +203,23 @@ olgm.herald.Feature.prototype.setVisible = function(value) {
...
+ * @private
+ * @return {ol.geom.Geometry} the feature's geometry
+ */
+olgm.herald.Feature.prototype.getGeometry_ = function() {
+  var geometry = this.feature_.getGeometry();
+  olgm.asserts.assert(
+      geometry !== undefined, 'Expected feature to have geometry');
+  return /** @type {ol.geom.Geometry} */ (geometry);
+};
+
```
Comment (adube) [inline src/olgm/herald/feature.js:216]: Suggestion (feel free to comment)

How about we move this method to `olgm.feature.getGeometry`, which would be located in `./src/feature.js`, as a utility method to get a geometry from a feature which is expected to have a geometry?

File: src/olgm/herald/feature.js [hunk 7]
```diff
@@ -207,13 +203,23 @@ olgm.herald.Feature.prototype.setVisible = function(value) {
...
 
 /**
  * @private
  */
 olgm.herald.Feature.prototype.handleGeometryChange_ = function() {
-  var geometry = this.feature_.getGeometry();
-  goog.asserts.assertInstanceof(geometry, ol.geom.Geometry);
+  var geometry = this.getGeometry_();
```
Comment (adube) [inline src/olgm/herald/feature.js:222]: Then use it here.

File: src/olgm/obj.js (hunk 1 above)
Comment (adube) [reply src/olgm/obj.js:14]: Nevermind this comment.

File: src/olgm/obj.js (hunk 4 above)
Comment (adube) [reply src/olgm/obj.js:38]: Nevermind.

**Review by adube — APPROVED** [review]
Excellent work.  Code is clean and everything makes sense.

I added a suggestion (see first comment), which isn't mandatory for this PR.

When the issue reported by Travis is fixed, we'll be good to merge this.

Thanks a lot.

**Comment by bampakoa** [comment]
@adube Shall we wait for this PR before upgrading to 4.5?

**Comment by adube** [comment]
@bampakoa Yes

**Comment by JamiesWhiteShirt** [comment]
@adube I am unable to reproduce the error Travis is reporting locally. `make test` runs just fine on Node 8 and 9.

The issue is that PhantomJS is silently unable to load the Google Maps API script on Travis. This is not reported by mocha-phantomjs-core because it is configured to ignore resource errors, seemingly because of some failing telemetry by the Google Maps API.

[As you can see in this experiment, Travis fails to load the Google Maps API script because of "SSL handshake failed"](https://travis-ci.org/mapgears/ol3-google-maps/jobs/326714466#L1359).

SSL and certificates isn't really my forté, so I don't know how to proceed.

**Comment by adube** [comment]
@JamiesWhiteShirt Thanks for your work on this.

Since there aren't that many tests anyway, I suggest to temporarely remove tests from being run by Travis until a fix for this is filled, as it doesn't seem to be related to the PR.

Let's try this on this patch, i.e. see in the `.travis.yml` file, replace

```
script: "make check"
```

by:

```
script: make lint && make build
```

Let me know what you think.

**Comment by adube** [comment]
I added a new issue about this: #201.

**Comment by JamiesWhiteShirt** [comment]
I did mostly as you asked. I assume you meant `make dist`.

**Comment by adube** [comment]
Oops, yep (sorry).  I'll make one final check and should be able to merge this really soon.

**Comment by adube** [comment]
Excellent work!  Thank you very much for this contribution.

**Comment by adube** [comment]
@bampakoa: we're ready for OL 4.5.0.

**Comment by bampakoa** [comment]
From OL 4.5.0 release notes:

> Please note that if you are using closure-util to build your OpenLayers based application, it is time to migrate to using the ol package and a module bundler like webpack

@adube Do we also need to incorporate **webpack** in the build process?

**Comment by adube** [comment]
Not for 4.5.

We'll wait for OL to make their 5.0 release before making a move.  I want to see how OL is going to get organized with its typing (currently with Closure Compiler and comments).  Webpack would be a logical choice.

**Pull request merged**